[Bug]: Cannot publish openzeppelin-stylus to crates.io

### What happened?

```
cargo publish -p openzeppelin-stylus --target wasm32-unknown-unknown  --dry-run
...
error[E0407]: method `stv_abi_packed_encoded_size` is not a member of trait `alloy_sol_types::private::SolTypeValue`
  --> src/utils/cryptography/ecdsa.rs:65:1
   |
65 | / sol! {
66 | |     /// Struct with callable data to the `ecrecover` precompile.
67 | |     #[allow(missing_docs)]
68 | |     struct EcRecoverData {
...  |
77 | |     }
78 | | }
   | | ^
   | | |
   | |_not a member of trait `alloy_sol_types::private::SolTypeValue`
   |   help: there is an associated function with a similar name: `stv_abi_encoded_size`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0438]: const `PACKED_ENCODED_SIZE` is not a member of trait `alloy_sol_types::SolType`
  --> src/utils/cryptography/ecdsa.rs:65:1
   |
65 | / sol! {
66 | |     /// Struct with callable data to the `ecrecover` precompile.
67 | |     #[allow(missing_docs)]
68 | |     struct EcRecoverData {
...  |
77 | |     }
78 | | }
   | |_^ not a member of trait `alloy_sol_types::SolType`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/access/control.rs:51:1
   |
51 | / sol! {
52 | |     /// Emitted when `new_admin_role` is set as `role`'s admin role, replacing
53 | |     /// `previous_admin_role`.
54 | |     ///
...  |
73 | |     event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender);
74 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/access/ownable.rs:19:1
   |
19 | / sol! {
20 | |     /// Emitted when ownership gets transferred between accounts.
21 | |     #[allow(missing_docs)]
22 | |     event OwnershipTransferred(address indexed previous_owner, address indexed new_owner);
23 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/token/erc20/mod.rs:18:1
   |
18 | / sol! {
19 | |     /// Emitted when `value` tokens are moved from one account (`from`) to
20 | |     /// another (`to`).
21 | |     ///
...  |
28 | |     event Approval(address indexed owner, address indexed spender, uint256 value);
29 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/token/erc721/extensions/consecutive.rs:71:1
   |
71 | / sol! {
72 | |     /// Emitted when the tokens from `from_token_id` to `to_token_id` are transferred from `from_address` to `to_address`.
73 | |     ///
74 | |     /// * `from_token_id` - First token being transferred.
...  |
84 | |     );
85 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/token/erc721/extensions/uri_storage.rs:11:1
   |
11 | / sol! {
12 | |     /// This event gets emitted when the metadata of a token is changed.
13 | |     ///
14 | |     /// The event comes from IERC4096.
...  |
23 | |     event BatchMetadataUpdate(uint256 from_token_id, uint256 to_token_id);
24 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/token/erc721/mod.rs:17:1
   |
17 | / sol! {
18 | |     /// Emitted when the `token_id` token is transferred from `from` to `to`.
19 | |     ///
20 | |     /// * `from` - Address from which the token will be transferred.
...  |
51 | |     event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
52 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0576]: cannot find method or associated constant `PACKED_ENCODED_SIZE` in trait `alloy_sol_types::SolType`
  --> src/utils/cryptography/ecdsa.rs:65:1
   |
65 | / sol! {
66 | |     /// Struct with callable data to the `ecrecover` precompile.
67 | |     #[allow(missing_docs)]
68 | |     struct EcRecoverData {
...  |
77 | |     }
78 | | }
   | |_^ not found in `alloy_sol_types::SolType`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0576]: cannot find method or associated constant `abi_packed_encoded_size` in trait `alloy_sol_types::SolType`
   --> src/utils/cryptography/ecdsa.rs:65:1
    |
65  | / sol! {
66  | |     /// Struct with callable data to the `ecrecover` precompile.
67  | |     #[allow(missing_docs)]
68  | |     struct EcRecoverData {
...   |
77  | |     }
78  | | }
    | |_^ help: an associated function with a similar name exists: `abi_encoded_size`
    |
   ::: /Users/bidzyyys/.cargo/registry/src/index.crates.io-6f17d22bba15001f/alloy-sol-types-0.7.6/src/types/ty.rs:130:5
    |
130 |       fn abi_encoded_size<E: ?Sized + SolTypeValue<Self>>(rust: &E) -> usize {
    |       ---------------------------------------------------------------------- similarly named associated function `abi_encoded_size` defined here
    |
    = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0405]: cannot find trait `IntoLogData` in module `alloy_sol_types::private`
  --> src/utils/pausable.rs:17:1
   |
17 | / sol! {
18 | |     /// Emitted when pause is triggered by `account`.
19 | |     #[allow(missing_docs)]
20 | |     event Paused(address account);
...  |
24 | |     event Unpaused(address account);
25 | | }
   | |_^ not found in `alloy_sol_types::private`
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0405, E0407, E0438, E0576.
For more information about an error, try `rustc --explain E0405`.
error: could not compile `openzeppelin-stylus` (lib) due to 11 previous errors
```

### platform

- [X] linux
- [ ] windows
- [X] macos

### Expected behavior

`openzeppelin-stylus` should be published to crates.io

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines

--- a/.github/workflows/check-links.yml
+++ b/.github/workflows/check-links.yml
@@ -15,16 +15,14 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
-      
+
       - name: Link Checker
         uses: lycheeverse/lychee-action@v1
         with:
           args: --no-progress './**/*.md'
           fail: true
-    
+
       - name: Run linkspector
         uses: umbrelladocs/action-linkspector@v1
         with:
           fail_on_error: true
-     
-        

--- a/.github/workflows/check-publish.yml
+++ b/.github/workflows/check-publish.yml
@@ -1,0 +1,41 @@
+name: check-publish
+# This workflow checks if the libraries can be published on crates.io.
+permissions:
+  contents: read
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+env:
+  CARGO_TERM_COLOR: always
+jobs:
+  check-publish:
+    name: Check publish on crates.io
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: set up rust
+        uses: dtolnay/rust-toolchain@master
+        id: toolchain
+        with:
+          target: wasm32-unknown-unknown
+          components: rust-src
+          toolchain: nightly-2024-01-01
+
+      - uses: Swatinem/rust-cache@v2
+
+      - name: check motsu-proc
+        run: cargo publish -p motsu-proc --dry-run
+
+      - name: check motsu
+        run: cargo publish -p motsu --dry-run
+
+      - name: check openzeppelin-crypto
+        run: cargo publish -p openzeppelin-crypto --target wasm32-unknown-unknown  --dry-run
+
+      - name: check openzeppelin-stylus
+        run: cargo publish -p openzeppelin-stylus --target wasm32-unknown-unknown  --dry-run

--- a/.github/workflows/check-publish.yml
+++ b/.github/workflows/check-publish.yml
@@ -28,14 +28,5 @@ jobs:
 
       - uses: Swatinem/rust-cache@v2
 
-      - name: check motsu-proc
-        run: cargo publish -p motsu-proc --dry-run
-
-      - name: check motsu
-        run: cargo publish -p motsu --dry-run
-
-      - name: check openzeppelin-crypto
-        run: cargo publish -p openzeppelin-crypto --target wasm32-unknown-unknown  --dry-run
-
       - name: check openzeppelin-stylus
         run: cargo publish -p openzeppelin-stylus --target wasm32-unknown-unknown  --dry-run

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -64,7 +64,7 @@ stylus-sdk = { version = "=0.6.0", default-features = false }
 stylus-proc = { version = "=0.6.0", default-features = false }
 mini-alloc = "0.4.2"
 
-alloy = { version = "0.1.4", features = [
+alloy = { version = "=0.1.4", features = [
   "contract",
   "network",
   "providers",
@@ -77,8 +77,8 @@ alloy = { version = "0.1.4", features = [
 # Even though `alloy` includes `alloy-primitives` and `alloy-sol-types` we need
 # to keep both versions for compatibility with the Stylus SDK. Once they start
 # using `alloy` we can remove these.
-alloy-primitives = { version = "0.7.6", default-features = false }
-alloy-sol-types = { version = "0.7.6", default-features = false }
+alloy-primitives = { version = "=0.7.6", default-features = false }
+alloy-sol-types = { version = "=0.7.6", default-features = false }
 
 const-hex = { version = "1.11.1", default-features = false }
 eyre = "0.6.8"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -103,7 +103,7 @@ quote = "1.0.35"
 
 # members
 openzeppelin-stylus = { path = "contracts" }
-openzeppelin-stylus-proc = { path = "contracts-proc" }
+openzeppelin-stylus-proc = { path = "contracts-proc", version = "0.1.0-rc" }
 openzeppelin-crypto = { path = "lib/crypto" }
 motsu = { path = "lib/motsu"}
 motsu-proc = { path = "lib/motsu-proc", version = "0.1.0" }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -79,6 +79,9 @@ alloy = { version = "=0.1.4", features = [
 # using `alloy` we can remove these.
 alloy-primitives = { version = "=0.7.6", default-features = false }
 alloy-sol-types = { version = "=0.7.6", default-features = false }
+alloy-sol-macro = { version = "=0.7.6", default-features = false }
+alloy-sol-macro-expander = { version = "=0.7.6", default-features = false }
+alloy-sol-macro-input = { version = "=0.7.6", default-features = false }
 
 const-hex = { version = "1.11.1", default-features = false }
 eyre = "0.6.8"

--- a/contracts-proc/Cargo.toml
+++ b/contracts-proc/Cargo.toml
@@ -1,12 +1,12 @@
 [package]
 name = "openzeppelin-stylus-proc"
 description = "Procedural macros for OpenZeppelin Stylus contracts"
-version.workspace = true
 authors.workspace = true
 edition.workspace = true
 license.workspace = true
 keywords.workspace = true
 repository.workspace = true
+version = "0.1.0"
 
 
 [dependencies]

--- a/contracts/Cargo.toml
+++ b/contracts/Cargo.toml
@@ -11,6 +11,9 @@ version = "0.1.0-rc"
 [dependencies]
 alloy-primitives.workspace = true
 alloy-sol-types.workspace = true
+alloy-sol-macro.workspace = true
+alloy-sol-macro-expander.workspace = true
+alloy-sol-macro-input.workspace = true
 stylus-sdk.workspace = true
 stylus-proc.workspace = true
 mini-alloc.workspace = true

--- a/docs/modules/ROOT/pages/crypto.adoc
+++ b/docs/modules/ROOT/pages/crypto.adoc
@@ -22,6 +22,6 @@ pub fn verify(&self, proof: Vec<B256>, root: B256, leaf: B256) -> bool {
 }
 ----
 
-Note that these functions use `keccak256` as the hashing algorithm, but our library also provides generic counterparts: https://docs.rs/openzeppelin-crypto/0.1.0-rc/openzeppelin_crypto/merkle/struct.Verifier.html#method.verify_with_builder[`verify_with_builer`] and https://docs.rs/openzeppelin-crypto/0.1.0-rc/openzeppelin_crypto/merkle/struct.Verifier.html#method.verify_multi_proof_with_builder[`verify_multi_proof_with_builder`].
+Note that these functions use `keccak256` as the hashing algorithm, but our library also provides generic counterparts: https://docs.rs/openzeppelin-crypto/0.1.0-rc/openzeppelin_crypto/merkle/struct.Verifier.html#method.verify_with_builder[`verify_with_builder`] and https://docs.rs/openzeppelin-crypto/0.1.0-rc/openzeppelin_crypto/merkle/struct.Verifier.html#method.verify_multi_proof_with_builder[`verify_multi_proof_with_builder`].
 
 We also provide an adapter https://docs.rs/openzeppelin-crypto/0.1.0-rc/openzeppelin_crypto/hash/index.html[`hash`] module to use your own hashers in conjunction with them that resembles Rust's standard library's API.

